NumBlocks, NumShards, NumSeats, HeightDelta and _seats

Kouprin opened the pull request:
1. Using `NumBlocks` and `BlockHeightDelta` in a proper way.
2. Correct naming in Epoch Manager.
3. Fixed https://github.com/nearprotocol/nearcore/issues/1855.
4. Unification of `BlockHeight` naming.

---

Set of apples in not an apple.
Size of set of apples is not an apple either.

Size of string is not a string.
Number of strings is not a string either.

`num_shards()` shouldn't return a `ShardId`.
Number of validators is not a `ValidatorId`.
Vector of _number_ of validators shouldn't be `Vec<ValidatorId>`.
Applying `.iter().sum()` to `Vec<ValidatorId>` shouldn't make any sense.

I expect mixing types leads to errors like https://github.com/nearprotocol/nearcore/issues/1855.

frol commented on core/primitives/src/types.rs:
```diff
@@ -27,6 +27,23 @@ pub type Balance = u128;
+pub type NumBlockProducers = u64;
+/// Number of seats of block producers in current group.
+// TODO MOO #1855: understand the difference between NumValidators, NumBlockProducers and NumSeats clearly
+pub type NumSeats = u64;
```
I completely support your vision about having explicit types instead of reusing some types just because they happen to have enough bytes to store the information!

However, I want to ask if we want to make it even more strict. `type` is just making an alias, and thus, one still can mess up and do `NumBlocks + NumShards - NumValidators`. Should we implement these types as single-value tuples with only a few explicit operations implemented?

Kouprin replied:
Personally, I'd prefer to keep it as agreement, not as strict rule.
Normally types should work as helpers for understanding the nature of the value. It should be okay to calculate `NumValidators + NumShards` for some specific reasons - those ones always appear.
I'm afraid that making types strict may lead to the following:
1. More casts, more methods; more code for compiler, not for developer.
2. "Temporary workarounds" to avoid using strict typing.
3. Difficulties with special cases.
4. Huge refactoring (once) which takes time and attention of code owners.

It will never too late to change to strict types. We can start with agreement easily and then see how does it work.

Kouprin commented on core/primitives/src/types.rs:
```diff
@@ -27,6 +27,23 @@ pub type Balance = u128;
+// Technically, blocks size is not equal to heights difference.
+// For example, 50 blocks don't imply heights from H to (H + 50).
+// It may lead to incorrect assumptions in future.
+// We can replace NumBlocks with HeightHorizon or something better.
```
I suppose `NumBlocks` and `HeightDelta` may work good enough.
For example, `epoch_length` means difference between heights when new epoch should start - use `HeightDelta`.
In initial state sync we want to request some number of blocks disregarding their heights - use `NumBlocks`.

bowenwang1996 commented on chain/epoch_manager/src/lib.rs:
```diff
@@ -876,7 +876,8 @@ impl EpochManager {
         // 121 - 118 = 3, hence the `last_finalized_height + 3`
         Ok((block_info.last_finalized_height + 3 >= estimated_next_epoch_start
-            || self.config.num_block_producers < 4)
+            // TODO MOO check if num_block_producer_seats meant here
```
yes this is a hack for tests.

bowenwang1996 commented on chain/epoch_manager/src/proposals.rs:
```diff
@@ -223,9 +226,10 @@ mod tests {
-                    num_block_producers: 6,
-                    block_producers_per_shard: vec![6, 2, 2, 2, 2],
-                    avg_hidden_validators_per_shard: vec![6, 2, 2, 2, 2],
+                    // TODO MOO total number of num_block_producer_seats is different to sum of _per_shard
```
doesn't have to be equal. There can be overlaps.

bowenwang1996 commented on core/primitives/src/types.rs:
```diff
@@ -27,6 +27,15 @@ pub type Balance = u128;
+/// Number of fishermen in current group.
+pub type NumFishermen = u64;
+/// Height delta.
+pub type HeightDelta = u64;
```
I think `HeightDelta` is confusing. `NumBlocks` might be better

Kouprin replied:
I check the code and I assume there should be both while we don't create blocks at each height.
`NumBlocks` means number of blocks in the group.
`HeightDelta` means difference of heights. It doesn't care how many blocks present in the group.

bowenwang1996 replied:
I don't think it makes sense to introduce a type for difference between two things of some other type. Otherwise we will just keep introducing new types and actually reduce the readability of the code.

Kouprin replied:
That's a huge difference between `NumBlocks` and `HeightDelta`.
Let's say, we have an epoch from height `H` to height `(H + epoch_length)`. `epoch_length` may be equal to 100, and the actual number of blocks between `H` and `(H + epoch_length)`, let's say, 31.

Let's see what's going on in `validate_chunk_transactions()`. We call `check_blocks_on_same_chain()` and then traverse `max_difference_in_blocks` which means taking real blocks, not height difference.

Let's see how we use `epoch_length`. We don't care how many blocks we have, we just care of `height` value. Heights difference is the criteria of a new epoch.

We cannot and shouldn't mix `NumBlocks` and `HeightDelta`. They have different meaning and describe different behavior.

Kouprin commented:
> Is it actually desirable to have for every concept A a corresponding type A' and for every container of A, a separate type A'' for its size?

It's discussable but I don't think so. The important thing is when A'' starts having its own meaning. For example, we do operate with number of seats and number of shards, that's why it's very logical to have separate types for them.

SkidanovAlex commented on chain/epoch_manager/src/proposals.rs:
```diff
@@ -144,21 +147,21 @@ pub fn proposals_to_epoch_info(
     let mut chunk_producers: Vec<Vec<ValidatorId>> = vec![];
     let mut last_index: u64 = 0;
-    for num_seats in epoch_config.block_producers_per_shard.iter() {
+    for num_seats_shard in epoch_config.num_block_producer_seats_per_shard.iter() {
```
`in_shard`

Kouprin replied:
++

SkidanovAlex commented on chain/epoch_manager/src/test_utils.rs:
```diff
@@ -74,23 +75,25 @@ pub fn epoch_info(
+    epoch_length: HeightDelta,
+    num_shards: NumShards,
+    num_block_producer_seats: NumSeats,
+    num_fishermen: NumFishermen,
```
There's no concept of `num_fishermen`. This should be `num_hidden_validators` (as it was), and the type should be just `NumSeats`

Kouprin replied:
++

SkidanovAlex commented on chain/epoch_manager/src/test_utils.rs:
```diff
@@ -74,23 +75,25 @@ pub fn epoch_info(
-        num_block_producers,
-        block_producers_per_shard: get_num_block_producers_per_shard(
+        num_block_producer_seats,
+        num_block_producer_seats_per_shard: get_num_seats_per_shard(
```
`num_block_producer_seats`, it returns speciflcally block producer seats, removing the words "block producer" from the name makes it unclear

Kouprin replied:
This functions splits Seats between Shards and doesn't know whose Seats are received in arguments.
It means we can reuse the function for Seats of any kind, not block producer only.
Do we really need to keep `get_num_block_producers_per_shard`?

Kouprin commented on near/src/config.rs:
```diff
@@ -122,10 +125,10 @@ pub const FISHERMEN_THRESHOLD: Balance = 10 * NEAR_BASE;
 
 /// Number of seats for block producers
-pub const NUM_BLOCK_PRODUCERS: ValidatorId = 50;
+pub const NUM_BLOCK_PRODUCER_SEATS: u64 = 50;
```
++

Kouprin commented on chain/client/src/test_utils.rs:
```diff
@@ -202,7 +202,7 @@ fn sample_binary(n: u64, k: u64) -> bool {
 ///                 for some groups will keep all the approvals (and test the fg invariants), and
 ///                 for some will drop 50% of the approvals.
-/// `epoch_length` - approximate number of heights per epoch
+/// `epoch_length` - approximate number of blocks per epoch
```
Discussed offline.

bowenwang1996 commented on chain/epoch_manager/src/test_utils.rs:
```diff
@@ -74,23 +81,28 @@ pub fn epoch_info(
+            num_block_producer_seats,
         ),
-        avg_hidden_validators_per_shard: (0..num_shards).map(|_| num_hidden_validators).collect(),
+        // TODO MOO that's weird because avg_hidden_validator_seats_per_shard will be summarize
```
This is for tests so it's fine.

Kouprin replied:
++

bowenwang1996 commented on core/primitives/src/types.rs:
```diff
@@ -18,15 +18,29 @@ pub type StorageUsage = u64;
+pub type HeightDelta = u64;
+
+#[derive(Default, BorshSerialize, BorshDeserialize, Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Seat {
```
It doesn't make sense to have this type for two reasons:
* Where this type is used it actually means an index in an array. I don't see why you want to wrap this type and makes code harder to read.
* Even if you want to wrap it, it is more idiomatic to have `pub struct Seat(ValidatorId)` so that the compiler will for sure optimizes away the wrapper.

Kouprin replied:
It's related to https://github.com/nearprotocol/nearcore/issues/1855. Let's discuss offline.

Kouprin replied:
Simplified to `pub type Seat = ValidatorId;`

Kouprin replied:
Removed `Seat` for now. We'll reopen this later if necessary.

bowenwang1996 commented on scripts/state/migrate-from-0.3.py:
```diff
@@ -77,8 +77,10 @@
 q['dynamic_resharding'] = False
-q['block_producers_per_shard'] = [50] * 8
-q['avg_fisherman_per_shard'] = [0] * 8
+# TODO please check this variable
```
what do you mean here?

Kouprin replied:
I don't actually know for what this script is. It called `migrate` which may mean that it needs some previous naming. I'm not sure, that's why I've added TODO to check with reviewers.

Kouprin commented:
> I don't think we need `Seat`. I looked at the places where you use seat and I think you mean validator_id instead. They are not the same and should not be confused with each other.

I removed `Seat` as you suggested. After that, `block_producers: Vec<ValidatorId>` may mean two separate things: block producer seats or list of block producers. This is ambiguous. I renamed all `block_producers` which related to seats to `block_producer_seats: Vec<ValidatorId>`.

It's okay to understand the object's type by name of variable. However I still don't understand why we shouldn't go to `Vec<Seat>` to resolve ambiguity by pointing to object nature directly.

Kouprin commented on chain/client/tests/challenges.rs:
```diff
@@ -540,17 +540,18 @@ fn test_verify_chunk_invalid_state_challenge() {
+                    152, 205, 174, 37, 238, 67, 244, 106, 68, 57, 151, 23, 90, 167, 66, 76, 171,
+                    30, 7, 228, 175, 99, 17, 113, 5, 94, 136, 200, 39, 136, 37, 110, 166, 241, 148,
+                    128, 55, 131, 173, 97, 98, 201, 68, 82, 244, 223, 70, 86, 137, 137, 2, 0, 0, 0,
+                    0, 0
```
the test relies on `runtime/near-vm-runner/tests/res/test_contract_rs.wasm` which was recompiled
that's why I've updated binary vectors in `test_verify_chunk_invalid_state_challenge`

bowenwang1996 commented:
> I removed `Seat` as you suggested. After that, `block_producers: Vec<ValidatorId>` may mean two separate things: block producer seats or list of block producers. This is ambiguous. I renamed all `block_producers` which related to seats to `block_producer_seats: Vec<ValidatorId>`.
> 
> It's okay to understand the object's type by name of variable. However I still don't understand why we shouldn't go to `Vec<Seat>` to resolve ambiguity by pointing to object nature directly.

Where does `Vec<ValidatorId>` mean a list of block producers?

Kouprin commented:
> > I removed `Seat` as you suggested. After that, `block_producers: Vec<ValidatorId>` may mean two separate things: block producer seats or list of block producers. This is ambiguous. I renamed all `block_producers` which related to seats to `block_producer_seats: Vec<ValidatorId>`.
> > It's okay to understand the object's type by name of variable. However I still don't understand why we shouldn't go to `Vec<Seat>` to resolve ambiguity by pointing to object nature directly.
> 
> Where does `Vec<ValidatorId>` mean a list of block producers?

In actual staging I found several HashMaps using `ValidatorId`: `fishermen_to_index`, `block_validator_tracker` / `chunk_validator_tracker`, `validator_to_index`. I suppose they are mapping from/to `ValidatorId` which means validator index, not seat. And `block_producers: Vec<ValidatorId>` means seats, not indices. If I'm right, it's ambiguity.

<img width="732" alt="Screenshot 2019-12-22 at 23 12 37" src="https://user-images.githubusercontent.com/5685493/71326751-96ec6280-2510-11ea-9bcd-bdf8cc8f0673.png">

Kouprin commented:
I also keep in mind that I may be wrong in assuming a difference between seats and indices. Maybe we don't even have such thing as `ValidatorId` and always mean seats. In this case we should rename `ValidatorId` to `ValidatorSeat`.

bowenwang1996 commented:
> And `block_producers: Vec<ValidatorId>` means seats, not indices. If I'm right, it's ambiguity.

But `ValidatorId` still means validator index in `block_producers`. There is no confusion as to what `ValidatorId` means. Renaming `block_producers` to `block_producer_seats` might also be confusing, in the other direction, because then people might think `ValidatorId` means seat, which is not the case.

bowenwang1996 commented:
> I also keep in mind that I may be wrong in assuming a difference between seats and indices. Maybe we don't even have such thing as `ValidatorId` and always mean seats. In this case we should rename `ValidatorId` to `ValidatorSeat`.

No. `ValidatorId` never means `Seat`, unless we have different understanding of what a seat is.

Kouprin commented:
At this point in our discussion I see the following:
1. We definitely have Validators which identified by ValidatorId.
2. We definitely have Seats and they're different from ValidatorIds.
3. I guess the issue https://github.com/nearprotocol/nearcore/issues/1855 is related to messing Seats and ValidatorIds.

I propose to name seat `Seat` and validator id `ValidatorId`.

bowenwang1996 commented:
Sure, but what's an example of something of type `Seat` that actually appears in the code?

Kouprin commented:
> Sure, but what's an example of something of type `Seat` that actually appears in the code?

I guess `EpochInfo::block_producers` is a vector of seats.

bowenwang1996 commented:
I see. I think we have different interpretation of `seat`. I consider the indices of `block_producers`, rather than its values, to be seats. In other words, I think we can think of `block_producers` to be of type `HashMap<Seat, ValidatorId>`, but there is no need for that so we just use a vector there.

Kouprin commented:
What do you suggest?

bowenwang1996 commented:
I suggest we keep things as is and do not introduce `Seat` unless we actually have something of type `Seat` in the code, but it seems that we have different interpretation of seat. Do you think my interpretation makes sense?

Kouprin commented:
Yes, we can go this way. How can we work on non-ambiguous understanding?
I expect it's easy to think the following concepts:
1. Real validators are exist and identified by ValidatorId.
2. Seats is a concept to describe the way of how we choose validators.

There may be 3 validators for 100 seats.
EpochConfig describes the number of seats by `num_block_producers`. EpochInfo stores the list of real validators in `validators`.
Does this understanding clear enough? Can we start with it?

If yes, the question is about `EpochInfo::block_producers: Vec<ValidatorId>`. What does it mean?

bowenwang1996 commented:
> If yes, the question is about `EpochInfo::block_producers: Vec<ValidatorId>`. What does it mean?

It means an array of seats on each of which is a validator identified by their validator id.

Kouprin commented:
> > If yes, the question is about `EpochInfo::block_producers: Vec<ValidatorId>`. What does it mean?
> 
> It means an array of seats on each of which is a validator identified by their validator id.

In `block_producers: Vec<ValidatorId>` definitely says nothing about seats. If it's an array of seats why not name it directly `Vec<Seat>`? If you don't like to introduce `Seat` as a type, should we rename `block_producers` to `block_producer_seats` as in current PR?

bowenwang1996 commented:
I don't know whether conflating seat with what the seats hold (validator id) is a good idea, but if you think it helps you understand the concept better, we can keep it this way. I would expect something of type `Vec<Seat>` to be an array whose values are indices that can be used to index into `block_producers`.

Kouprin commented:
Is it correct to say that value of `Seat` is `ValidatorId`? If we have 3 validators and 100 seats, can I say that seat #52 is occupied by validator #1?

Kouprin commented:
I'm trying to make my understanding more visual to specify and correct it.
---

There are Isaac, Maggy and Samson, and they are validators.

<img width="160" alt="Screenshot 2019-12-23 at 12 25 50" src="https://user-images.githubusercontent.com/5685493/71350002-5c6ede00-2581-11ea-9bf8-178a3c6909ac.png">

They are stored at `EpochInfo::validators: Vec<ValidatorStake>`.

<img width="237" alt="Screenshot 2019-12-23 at 12 28 30" src="https://user-images.githubusercontent.com/5685493/71351190-1cf5c100-2584-11ea-8b2a-b0db8df4234d.png">

According to the PR, there is a parameter `EpochConfing::num_block_producer_seats: NumSeats` which says how many seats we have in total. The validators came to occupy seats in function `proposals_to_epoch_info()`.  Isaac got 3 seats, Maggy got 2 seats and Samson got 1 seat.

<img width="593" alt="Screenshot 2019-12-23 at 12 25 46" src="https://user-images.githubusercontent.com/5685493/71350134-ae176880-2581-11ea-9670-30a6fd88101b.png">

We need to keep the assignment. That's why we have `EpochInfo::block_producer_seats: Vec<Seats>`.

<img width="263" alt="Screenshot 2019-12-23 at 12 28 26" src="https://user-images.githubusercontent.com/5685493/71350631-b91ec880-2582-11ea-85d4-ed2d4a544e8c.png">

Seat is simple. It only says who occupies it. So, it stores `ValidatorId`. 

At some moment in time we want to ask: who sits at the seat #4? The asking is happening in `EpochManager::get_block_producer_info()` function. In this case `SeatId = 4` but we never use it explicitly. We ask `EpochInfo::block_producer_seats[4]` and it says: there is `ValidatorId = 1`.

Okay! Now we check `EpochInfo::validators: Vec<ValidatorStake>` with `ValidatorId = 1` and get Maggy!

Hope it helps.

bowenwang1996 commented:
> Is it correct to say that value of `Seat` is `ValidatorId`? If we have 3 validators and 100 seats, can I say that seat #52 is occupied by validator #1?

Yes. I am fine with `block_producers_settlement` as well. Feel free to choose the one that you prefer.

SkidanovAlex commented on near/src/runtime.rs:
```diff
@@ -573,17 +579,18 @@ impl RuntimeAdapter for NightshadeRuntime {
         let epoch_id = epoch_manager.get_epoch_id_from_prev_block(parent_hash).unwrap();
-        if let Ok(block_producers) = epoch_manager.get_all_block_producers(&epoch_id, &parent_hash)
+        if let Ok(settlement) =
+            epoch_manager.get_all_block_producers_settlement(&epoch_id, &parent_hash)
```
The length of `settlement` is constant throughout live of the system.
We should return that length from the config, not compute an unnecessary constant size array each time and then take its length.
Also that will remove `parent_hash` from the arguments of `num_total_parts`, which is very desirable.

Kouprin replied:
++

Kouprin commented:
> > It applies Russell's type theory explicitly.
> 
> How is Russell's type theory related to variable naming?

Not using the same type for objects and groups of objects, a very local and pragmatic application.

> > Further reading: https://en.wikipedia.org/wiki/Type_theory
> 
> How is this article relevant to variable naming?

It's related to the idea of "not using the same type for objects and groups of objects".

> What is the rationale behind the renaming block index to block height? Aren't those exact synonyms? This change is too invasive, it modifies smart contract API that we have frozen several months ago. Only extensions are allowed at this point unless we find a bug which is not the case here.

The point is to be more concrete and precise in naming object types. Using heights and indices with equal meaning creates some mess. `block_height` may easily transform to `index`, however there are some places where we use indices with other meaning, i.e. `receipt_index`. As I started renaming, I decided to clean the part related to block heights too.

> smart contract API that we have frozen several months ago

Sorry for that. Could you please more specific and suggest a solution here? I see the following options:

1. Ignore the mess. Use heights and indices as synonyms and revert the change related to renaming indices to heights - the worst one because we keep more room for mess.
2. Rename all `heights` to `indices` and stop using `heights`. I prefer `heights` because they are more block-related than indices. We may find indices somewhere else but not heights.
3. Rename `indices` to `heights` everywhere except contracts.
4. Add `block_height` to the contract and start using it instead of `block_index` - as we are allowed to add extensions.
5. Rename `heights` to `indices` everywhere - as I see, it doesn't work for you.

Kouprin commented on runtime/near-vm-runner/src/imports.rs:
```diff
@@ -52,7 +52,7 @@ wrapped_imports! {
     signer_account_pk<[register_id: u64] -> []>,
     predecessor_account_id<[register_id: u64] -> []>,
     input<[register_id: u64] -> []>,
-    block_index<[] -> [u64]>,
```
Answered below. Let's be more concrete and discuss about solution.

nearmax commented:
> Not using the same type for objects and groups of objects, a very local and pragmatic application.

Guiding the code by best software development practices is better than guiding it by some mathematical theories. If you look through Google style guides on Java and C++ the most guidance you will see there on how to name stuff is basically of two types: 1) use plurals for collections and singular for everything else; 2) expand non-common abbreviations. This aligns with your PR so it is more like a general comment.

> however there are some places where we use indices with other meaning, i.e. receipt_index.

By the same argument, we should rename "receipt index" to "receipt height" (or whatever other synonym, like "receipt depth". Especially since we are the one who came up with the receipts we can make up any terminology there) everywhere because we have a "transaction index" in some places and people might incorrectly interpret the meaning of "index" in "receipt index".

> Using heights and indices with equal meaning creates some mess. 

I think the assessment that using both "block index" and "block height" is of huge importance is subjective. Frequently people write a code that uses "element number" and "element index" interchangeably to refer to a position of an element in a collection.

> I prefer heights because they are more block-related than indices.

To me, block height, block index, and block number are completely synonymous. Quick google search confirms that. It is not like blockchain is a decades-old academic field with established terminology so the fact that "block height" is more used than "block index" is of small importance.

> Rename all heights to indices and stop using heights. I prefer heights because they are more block-related than indices. We may find indices somewhere else but not heights.

This sounds reasonable. I don't see a reason why blockchain folks are using non-common names for common things. By doing that we might bring a bit of order to the world of blockchain terminology, otherwise, people will continue coming up with creative names to describe the position of an element in a linear order. "block height", "receipt depth", "transaction nonce" are much more confusing to me than "block index", "receipt index", "transaction index" etc.

> Add block_height to the contract and start using it instead of block_index - as we are allowed to add extensions.

It will be pretty bad if our first stable version of the blockchain code will already have a legacy API in it.

> Rename heights to indices everywhere - as I see, it doesn't work for you.

Let me clarify why this is actually a problem. If we change this API we would have to hard-fork, update most of the Applayer components, and redeploy them simultaneously with the network. As we learned recently this is not an easy trick and requires up to a week of synchronization between the people, also the clients will start writing us asking why their stuff stopped working and what should they do with it. We already have two hard-forks in the next month, adding another hard-fork because we prefer "block height" over "block index" sounds excessive, we might end up with a hard fork almost every week, which is very disrupting, since preparing hard-fork takes about a week (in terms of communicating with person behind every involved component).

Kouprin commented:
@nearmax my main request to you is to come up with the solution. I agree with some things you noticed above and I don't want to get into what-exactly-I-mean arguing for some things like renaming "receipt index" to "receipt height". I'm OK with most solutions except those which makes things worse and I clearly see why it's happened.

Once again. What is the best solution you suggest?

nearmax commented:
> @nearmax my main request to you is to come up with the solution. I agree with some things you noticed above and I don't want to get into what-exactly-I-mean arguing for some things like renaming "receipt index" to "receipt height". I'm OK with most solutions except those which makes things worse and I clearly see why it's happened.
> 
> Once again. What is the best solution you suggest?

As I suggested in the previous comment we rename block height to block index everywhere.

nearmax commented:
On a second look, based on the commit you have just submitted, renaming every block height to every block index still breaks everything in the Applayer and will require a hard fork anyway. It also seems to be touching significantly more code than renaming block index to block height. So since we cannot avoid the hard fork anyway, there is no real difference which one to use, and since block height is more commonly used I am actually leaning towards it. Let's poll people in Slack on what they prefer.

Kouprin commented:
> As I suggested in the previous comment we rename block height to block index everywhere.

OK, I changed block height to block index. Subjectively, I found that:

1. Block height sounds more about block. Block index messes with other indices.
2. Because of (1), we should use two words to describe the concept: "block index", not just one word "height". It may lead to some slang abbreviations or confusion (in case of speaking about indices a lot without adding "block" every time).
3. Some weird names appeared, like `last_chunk_block_index_included`.

"Height" doesn't have such bad side-effects.

bowenwang1996 commented:
In the last retro meeting about the yotcoNEAR update I think we all agreed that unless there is a very good reason we *should not* do hard fork. This renaming is not a good enough reason to me.

nearmax commented:
Are you suggesting that inconsistent naming in node API is not good enough reason?

bowenwang1996 commented:
I think it's better to introduce new API and deprecate old ones.

nearmax commented:
> I think it's better to introduce new API and deprecate old ones.

AFAIU that's one of the plans we were discussing in slack.

nearmax commented:
> I think it's better to introduce new API and deprecate old ones.

We should not be launching our mainnet with already deprecated APIs. So we should get rid of block index in smart contract API, the problem is, the moment we do it some of the currently deployed smart contracts will come invalid because they will fail to execute. This means the old blocks will become invalid, and the only way to run this network would be to do the state dump, which is a hard fork. It does not matter in what order we roll out the API changes it is going to be a hard fork anyway.

nearmax commented on runtime/near-vm-logic/tests/test_context.rs:
```diff
@@ -11,7 +11,7 @@ pub fn create_context() -> VMContext {
         predecessor_account_id: "carol".to_owned(),
         input: vec![0, 1, 2, 3, 5],
-        block_index: 10,
+        height: 10,
```
Word "block" should not be omitted. Block height is more descriptive. We are not omitting word block in block timestamp, and we are not omitting "attached" in the "attached deposit" either.

Kouprin replied:
I checked manually and found that for many cases using `block_height` instead `height` is useless. We don't use `block_height_created` and `block_height_included`, we use `height_created` and `height_included`. We use `height` in `BlockHeaderInnerLite`. We use `get_block_hash_by_height()` instead of `get_block_hash_by_block_height()`.

If you still want to add `block_` to `height` somewhere, please point all the places where you think it's necessary.

nearmax replied:
That's because chain and runtime have different contexts. When some developer is working with chain and sees "height" they obviously know that this height is related to block height. The runtime does not have this context, because it is an engine that executes stuff (it does not need blockchain to run) so when someone encounters "height" inside the runtime code they don't know what it relates to. I suggest we use "block_height" everywhere in the `nearcore/runtime` and `nearcore/core` (unless the primitives like `block.rs`). The reason why we would keep "block" in `nearcore/core` is because it does not have context of the "chain" either.

Kouprin replied:
According to your comment, I've updated some "height"s to "block_height"s where it makes sense.

nearmax commented on runtime/near-vm-runner/src/imports.rs:
```diff
@@ -52,7 +52,7 @@ wrapped_imports! {
     predecessor_account_id<[register_id: u64] -> []>,
     input<[register_id: u64] -> []>,
-    block_index<[] -> [u64]>,
+    height<[] -> [u64]>,
```
This PR cannot be submitted on its own because it will break a bunch of stuff. It should be accompanied by PRs to nearlib (maybe nearshell), explorer, near-bindgen, near-runtime-ts, and it should include a procedure of hard fork: dumping the state and restarting the nodes+applayer as we discussed in slack. CC @vgrichina @bowenwang1996

Kouprin replied:
Sure. Can we work on top of this PR?

nearmax replied:
It depends on what do you mean by working on top of this PR. We can only merge code into staging if it becomes a part of the new release. This code cannot be merged into staging without merging corresponding fixes to other repositories.

Kouprin replied:
Yes, the question is to merge into staging or creating a new branch to continue development on top of this commit. I accept any of these decisions. We shouldn't create a huge diff between this PR and actual staging by delaying if possible.

nearmax replied:
Everyone else will continue developing on the top of staging which is also a branch that we use for the releases (approx once a week we merge staging into master and cut a release). We cannot ask everyone to continue working on top of branch other than staging.

I think there is only one way to merge this PR is prepare PRs in other repos and simultaneously do the following:
* merge all PRs;
* dump current network state and restart it with new version -- hard fork.

Kouprin replied:
> We cannot ask everyone to continue working on top of branch other than staging.

Why not?

nearmax replied:
Because then it will become a staging branch but with an extra step of inconveniencing everyone by asking them to switch to a different branch from staging.

Merging this PR to staging is equivalent to not merging it and asking everyone to work off this branch.

Kouprin commented:
CI is broken:
<img width="856" alt="Screenshot 2019-12-27 at 20 07 01" src="https://user-images.githubusercontent.com/5685493/71529328-7c393580-28e4-11ea-91be-211adab55ec8.png">
Example: https://gitlab.com/near-protocol/nearcore/-/jobs/389881785

nearmax commented:
> CI is broken:
> <img alt="Screenshot 2019-12-27 at 20 07 01" width="856" src="https://user-images.githubusercontent.com/5685493/71529328-7c393580-28e4-11ea-91be-211adab55ec8.png">
> Example: https://gitlab.com/near-protocol/nearcore/-/jobs/389881785

Fixing it is on you unless it is a CI bug then you should write @ailisp .

Kouprin commented:
> Fixing it is on you unless it is a CI bug then you should write @ailisp .

"CI is broken" doesn't mean "tests are failed".

Kouprin commented:
@nearmax which changes you're requesting?

nearmax commented:
> > Fixing it is on you unless it is a CI bug then you should write @ailisp .
> 
> "CI is broken" doesn't mean "tests are failed".
> 
> Created #1902.

Our CI checks more than just tests. In fact the CI failure tell exactly what went wrong -- we enforce code formatting through cargo fmt, please fix it.

nearmax commented:
> @nearmax which changes you're requesting?

To avoid the hardfork any renaming of the struct names or struct field names or enum names or enum variant names in `core/primitives` should be undone, plus changes to `runtime/near-vm-runner/src/imports.rs` and `runtime/near-vm-logic/src/logic.rs` interface. If you want to avoid restarting the testnet anything that renames fields in the configs (there are several of them) should be undone too. If you want to avoid fixing `near-bindgen` you should undo renaming of fields and methods in `runtime/*` crates.

Kouprin commented:
How difficult to restart testnet?

nearmax commented:
> How difficult to restart testnet?

AFAIU it is not really difficult -- kill and restart processes on several gcloud machines. The problem is that testnet will be down for some time so this should be carefully communicated with whatever events, talks, and stakewars that we are running. Plus there is always a chance that something might go wrong so be ready to spend several hours doing urgent fixes across repos (I cannot tell you what might go wrong, I am saying there is always a chance of that). Plus before you do all that you would need to dry-run the network on separate nodes, preferably with Applayer pointing to it to make sure we are not breaking it. CC @vgrichina Also, please read everything here: https://www.notion.so/nearprotocol/Releases-2373f1847cef4b74b960586a32a6b8de

Kouprin commented:
It's not necessary to be forced. It's OK to apply the changes to the next release date. Does it block merging to staging?

Kouprin commented:
> To avoid the hardfork any renaming of the struct names or struct field names or enum names or enum variant names in core/primitives should be undone, plus changes to runtime/near-vm-runner/src/imports.rs and runtime/near-vm-logic/src/logic.rs interface. If you want to avoid restarting the testnet anything that renames fields in the configs (there are several of them) should be undone too. If you want to avoid fixing near-bindgen you should undo renaming of fields and methods in runtime/* crates.

According to this comment, I left all the names unchanged in:
- /core/primitives - all struct names or struct field names or enum names or enum variant names are undone
- /runtime/* - all struct names or struct field names or enum names or enum variant names or methods names or functions names are undone

I also added comments to all places where renaming should be added later "TODO https://github.com/nearprotocol/nearcore/issues/1903".
